### PR TITLE
M154 Position Auto-Report

### PR DESCRIPTION
**Pull request by shitcreek**
Add auto position reporting with `M154 S`
e.g., `M154 S2` - Report position every 2 seconds

Tested and working

Requested by @G2Barbour
https://github.com/MarlinFirmware/Marlin/issues/18247

### CODE REVIEW

**Comment by PrintEngineering** [comment]
Good to see the feature added! I do have one update on the functionality and recommend the following change in temperature.cpp: 

Current code:

        if(gcode.autoreport_position){
          SERIAL_CHAR(' ');
          report_current_position_projected();}
        else
          SERIAL_EOL();


New Code:
        SERIAL_EOL();
       if(gcode.autoreport_position) report_current_position_projected();


The reason for this change is that it is currently beyond the ability of Octoprint to parse temperatures from the combined long line. Splitting the output into two lines appears to external hosts functionally equivalent to an M105 followed by an M114 .

**Comment by thisiskeithb** [comment]
Doesn't `M114_DETAIL` also report XYZE position?

**Comment by PrintEngineering** [comment]
> Doesn't `M114_DETAIL` also report XYZE position?

Indeed. but not *autoreport*

**Comment by thinkyhead** [comment]
At the moment this is merely a hack proof-of-concept.
- Adding position reporting on top of Temperature reporting is the hack part.
- No host software knows how to parse the output from this PR.

What you need is a new option that enables the feature, plus a separate timer, plus a separate interval, plus a new G-code to enable the feature and set options. It should probably be implemented as an extension to `M105`.

**Comment by PrintEngineering** [comment]
Though BTT TFT firmware is able to parse it, as my custom branch was able to do without modification, I noticed a problem with Octoprint and this is what I had in mind so that it could be parsed by any firmware, this is my comment from *earlier*:

"Good to see the feature added! I do have one update on the functionality and recommend the following change in temperature.cpp:

Current code:

    if(gcode.autoreport_position){
      SERIAL_CHAR(' ');
      report_current_position_projected();}
    else
      SERIAL_EOL();
New Code:
SERIAL_EOL();
if(gcode.autoreport_position) report_current_position_projected();

The reason for this change is that it is currently beyond the ability of Octoprint to parse temperatures from the combined long line. Splitting the output into two lines appears to external hosts functionally equivalent to an M105 followed by an M114 ."

**Comment by PrintEngineering** [comment]
But ideally it would be a separate implementation, as you said. I hope it is added to M105, that would be great. I'm very new to poking around in Marlin code, so this was the best I could pull off without making a huge project about it.

**Comment by thinkyhead** [comment]
> it is currently beyond the ability of Octoprint to parse temperatures from the combined long line

Exactly.

Now, if this is for the benefit of some display that acts like a host, great. But truly, hosts don't need this. Marlin already reports all changes in position that the host needs to know about. And there's no good reason that hosts should need a continuous stream of (outdated) position information.

**Comment by PrintEngineering** [comment]
I have an Artillery Sidewinder and our community has unusual circumstances to deal with. We have a TFT that does not support Marlin Mode - no encoder. I have been implementing M600 with Host Prompt Support and Host Action Commands. With this method, and Trojan horsing Marlin by defining an LCD controller, I have been able to successfully use M600 and retain use of the filament runout sensor from the TFT, where it was originally installed, even after moving it to the main board. This gives us the ability to use the filament runout sensor with Octoprint and also use M600 from both Octoprint and while printing from the TFT. The problem with this new configuration that this new polling feature solves is that when Octoprint is acting as the main host, the TFT has no idea what the current Z position is. The alternative would be to have octoprint echo that information automatically in a similar way. I'm not sure what would be easier on Marlins ability to keep up with and process the simultaneously incoming gcode though. So far this does seem to work.

**Comment by shitcreek** [comment]
Since M105 gets the information from the host at `some point in the future` wouldn't the information be even more outdated than M155?
I'm going to dig deeper into this.

**Comment by PrintEngineering** [comment]
I'm not sure about the relationship to M105 and how current the data is. I'm brand new to Marlin. But the more I think about it, having it combined with M155 instead of moving it to a separate command is far more elegant than having to send 2 separate commands. If you want autoreporting, it makes sense to get it all in one place.

It could, potentially for some people, be beneficial to add another timer. So if 'P' has no value with it, the 'S' value defines the timescale for the 'P' as well, just like it works with the current functionality. But if you send M155 P2 S1 you can have each on a different timer -  2 seconds and 1 second respectively. 

But personally, I prefer it exactly the way it is. It will never require less resources than the way it is implemented now and I think having them reported together is clean, effective, and compatible with any host now that the output has been split to 2 lines.

**Comment by shitcreek** [comment]
What's the consensus on this? should i close this PR?

**Comment by thinkyhead** [comment]
I've reworked this code to make it more consistent with existing patterns. The interval is now set using `M154 S` instead of `M155 S`. The updated code uses the handy `AutoReporter` template class so it can have its very own timing interval separate from Temperature and SD Status auto-reporters, but it will of course pause along with the other auto-reporters as needed.

I've also filled in the various pieces that go into completing a major new feature:
- Added a new configuration option `AUTO_REPORT_POSITION` to `Configuration_adv.h`.
- Updated all `Configuration_adv.h` files over at `Configurations/import-2.0.x` with the new option.
- Added the `gcode/host/M154.cpp` to the `platformio.ini` exclusion filters.
- Added `AUTO_REPORT_POSITION` to the `ini/features.ini` inclusion filters.
- Added `AUTO_REPORT_POSITION` to (at least one) CI build test (`mega2560`).
- Added a new capability — `AUTOREPORT_POS` — to the `M115` report.
- Documented the new `M154` G-code on the RepRap wiki ([here](https://reprap.org/wiki/G-code#M154:_Auto_Report_Position)).
- Documented the new `AUTOREPORT_POS` capability on the RepRap wiki ([here](https://reprap.org/wiki/Firmware_Capabilities_Protocol#Cap:AUTOREPORT_POS)).

**Comment by thinkyhead** [comment]
Congratulations! and thank you for your patience. It was hard to get to the backlog in the long run-up to v2.0.8, but it turns out this was straightforward to get through cribbing everything from the temperature report. 🤓

File: Marlin/src/gcode/gcode.cpp [hunk 1]
```diff
@@ -561,6 +561,10 @@ void GcodeSuite::process_parsed_command(const bool no_ok/*=false*/) {
         case 193: M193(); break;                                  // M193: Wait for cooler temperature to reach target
       #endif
 
+      #if ENABLED(AUTO_REPORT_POSITION)
+        case 154: M154(); break;                                  // M155: Set position auto-report interval
```
Comment (kisslorand) [inline Marlin/src/gcode/gcode.cpp:565]: "// __M155__: Set position auto-report interval"
Typo: "M155" instead of "M154".

**Comment by G2barbour** [comment]
Wow! Excellent! Glad to see it come to life! I haven't been messing with my 3D printer all year but when. I have the time I look forward to testing it out! 

Thank you!

**Comment by sarvenn** [comment]
Hi,
M154 command explanation is missing on the Marlin FW webpage. Could anyone add that there? @thinkyhead @thisiskeithb 
https://marlinfw.org/meta/gcode/

**Pull request merged**